File: 19 risk - external tools, bracket fixed/cpp/bracket_fixed.cpp
#include "tse_helpers.hpp"

#include <cstdio>
#include <string>
#include <vector>

int main()
{
	tse::setLogLevel(tse::LogLevel::none);

	tse::Account account {"RiskManagementExternalTools BracketFixed", tse::StorageRegime::mem};
	std::vector<tse::OhlcvTick> const ticks
		{tse::Account::loadOhlcvCsv(helpers::dataPath("WTI_OHLCVminute_sept2016.csv"))};

	tse::Market const market {account.createMarket("OHLCV", tse::MdType::ohlcv)};
	account.createSimulator("Sim", helpers::simulatorConfig(), 10, -1);
	account.addContract("WTI", 1, tse::Instrument::future, tse::Underlying::commodity, tse::Venue::undefined, 100000);

	account.addInputOhlcv("WTI SMA fast", 3, tse::Duration::days, helpers::makeSma(3), market, {"WTI"});
	account.addInputOhlcv("WTI SMA slow", 10, tse::Duration::days, helpers::makeSma(10), market, {"WTI"});
	account.addPatternCrossover("PatternToLong", tse::Duration::days, {"WTI SMA fast", "WTI SMA slow"}, tse::Cmp::ge);

	// Buy 580 lots of crude as soon as the three-day average climbs above the ten-day one.
	tse::LegDescriptor const entry
	{
		"WTI",
		tse::QuantityMode::fixed,
		580.0,
		tse::PriceType::market,
		45.0,
		0.0,
		0.0,
		tse::TxnType::enter,
		tse::Side::long_,
		tse::Side::neutral,
		tse::Tif::day,
		tse::Priority::replaceable
	};
	// Here the risk is not recomputed on every tick. The stop 2% below the fill and the target 3%
	// above it go out with the entry and rest at the venue, so they guard the position even if
	// this program stops running.
	tse::VenueRiskSpec const risk
	{
		tse::VenueRiskLeg {true, 0, 0.02},
		tse::VenueRiskLeg {true, 0, 0.03},
		tse::Tif::gtc
	};
	account.addRuleBracket("Bracket", entry, risk, "PatternToLong");
	account.addRobot("Bracket robot", {"Bracket"});
	account.start("Bracket robot");

	for (tse::OhlcvTick const& tick : ticks) {
		market.pushOhlcv("WTI", tick);
	}

	tse::Summary const summary {account.getSummary()};
	std::printf
	(
		"bracket fixed: trades=%lld netProfit=%.4f grossProfit=%.4f grossLoss=%.4f profitFactor=%.6f maxDrawdown=%.4f\n",
		static_cast<long long>(summary.totalNumberOfTrades),
		summary.totalNetProfit,
		summary.grossProfit,
		summary.grossLoss,
		summary.profitFactor,
		summary.maxDrawdown
	);
	return 0;
}

File: 19 risk - external tools, bracket fixed/python/bracket_fixed.py
import os
import sys

sys.path.insert(0, os.path.join(os.path.dirname(__file__), "..", "..", "helpers", "python"))
import tse_helpers as H

tse = H.tse


def main():
    account = H.account("RiskManagementExternalTools BracketFixed", tse.StorageRegime.Mem, tse.LogLevel.Off)
    ticks = account.load_ohlcv_csv(H.data_path("WTI_OHLCVminute_sept2016.csv"))

    market = account.create_market("OHLCV", tse.MdType.Ohlcv)
    account.create_simulator("Sim", H.simulator_options(), 10, -1)
    account.add_contract("WTI", 1, tse.Instrument.Future, tse.Underlying.Commodity, tse.Venue.Undefined, 100000)

    account.add_input_ohlcv("WTI SMA fast", 3, tse.Duration.Days, H.make_sma(3), market, ["WTI"])
    account.add_input_ohlcv("WTI SMA slow", 10, tse.Duration.Days, H.make_sma(10), market, ["WTI"])
    account.add_pattern_crossover("PatternToLong", tse.Duration.Days, ["WTI SMA fast", "WTI SMA slow"], tse.Cmp.Ge)

    # Buy 580 lots of crude as soon as the three-day average climbs above the ten-day one.
    entry = tse.make_leg_descriptor(
        "WTI",
        tse.Quantity.Fixed,
        580.0,
        tse.Price.Market,
        45.0,
        0.0,
        0.0,
        tse.TxnType.Enter,
        tse.Side.Long,
        tse.Side.Neutral,
        tse.Tif.Day,
        tse.Priority.Replaceable,
    )
    # Here the risk is not recomputed on every tick. The stop 2% below the fill and the target 3%
    # above it go out with the entry and rest at the venue, so they guard the position even if
    # this program stops running.
    risk = tse.make_venue_risk_spec(
        tse.make_venue_risk_leg(True, 0, 0.02),
        tse.make_venue_risk_leg(True, 0, 0.03),
        tse.Tif.Gtc,
    )
    account.add_rule_bracket("Bracket", entry, risk, "PatternToLong")
    account.add_robot("Bracket robot", ["Bracket"])
    account.start("Bracket robot")

    for tick in ticks:
        market.push_ohlcv_by_name("WTI", tick)

    summary = account.get_summary()
    account.close()

    print("bracket fixed: trades={} netProfit={:.4f} grossProfit={:.4f} grossLoss={:.4f} profitFactor={:.6f} maxDrawdown={:.4f}".format(
        summary.totalNumberOfTrades, summary.totalNetProfit, summary.grossProfit,
        summary.grossLoss, summary.profitFactor, summary.maxDrawdown))
    return 0


if __name__ == "__main__":
    sys.exit(main())
